pkgs/development/tools/parsing/tree-sitter/default.nix

```diff
@@ -105,6 +108,48 @@ let
 
   allGrammars = builtins.attrValues builtGrammars;
 
+
+  configWithGrammarSources = grammarFn:
+    let
+      # Since names aren't readable from the imported object, map each
+      # object to { "tree-sitter-xyz": "tree-sitter-xyz" }
+      selectedNames = grammarFn (builtins.mapAttrs (k: v: k) grammarSources);
+      selectedGrammars = builtins.listToAttrs (builtins.map (x: { name = x; value = grammarSources.${x}; }) selectedNames);
```

Comment:
not super proud of this, but the name of the imported grammar isn't available in the existing derivations, so keeping it around and matching it back works, and keeps the same familiar syntax of `withX(p: [...])` while still being able to linkFarm with the appropriate name). I'm opened to improvements suggestions.

Reply:
instead you could add an attribute to sources: `sources = mapAttrs (k: v: v // { grammarName = k }) grammarSources`